ethergo/submitter/submitter.go

```diff
@@ -346,119 +346,211 @@ func (t *txSubmitterImpl) txTypeForChain(chainID *big.Int) (txType uint8) {
+// If gas values exceed the configured max, an error will be returned.
 func (t *txSubmitterImpl) setGasPrice(ctx context.Context, client client.EVM,
 	transactor *bind.TransactOpts, bigChainID *big.Int, prevTx *types.Transaction) (err error) {
 	ctx, span := t.metrics.Tracer().Start(ctx, "submitter.setGasPrice")
 
 	chainID := int(bigChainID.Uint64())
-	maxPrice := t.config.GetMaxGasPrice(chainID)
+	useDynamic := t.config.SupportsEIP1559(chainID)
+	shouldBump := prevTx != nil
 
 	defer func() {
-		if transactor.GasPrice != nil && maxPrice.Cmp(transactor.GasPrice) < 0 {
-			transactor.GasPrice = maxPrice
+		span.SetAttributes(
+			attribute.Int(metrics.ChainID, chainID),
+			attribute.Bool("use_dynamic", useDynamic),
+			attribute.Bool("should_bump", shouldBump),
+			attribute.String("gas_price", bigPtrToString(transactor.GasPrice)),
+			attribute.String("gas_fee_cap", bigPtrToString(transactor.GasFeeCap)),
+			attribute.String("gas_tip_cap", bigPtrToString(transactor.GasTipCap)),
+		)
+		metrics.EndSpanWithErr(span, err)
+	}()
+
+	t.populateGasFromPrevTx(ctx, transactor, prevTx, useDynamic)
+	t.applyGasFloor(ctx, transactor, chainID, useDynamic, shouldBump)
+
+	err = t.applyGasFromOracle(ctx, transactor, client, useDynamic)
+	if err != nil {
+		return fmt.Errorf("could not populate gas from oracle: %w", err)
+	}
+
+	err = t.applyGasCeil(ctx, transactor, chainID, useDynamic)
+	if err != nil {
+		return fmt.Errorf("could not apply gas ceil: %w", err)
+	}
+	return nil
+}
+
+// populateGasFromPrevTx populates the gas fields from the previous transaction.
+// Note that in the event of a tx type mismatch, gasFeeCap is copied to gasPrice,
+// and gasPrice is copied to both gasFeeCap and gasTipCap in the opposite scenario.
+//
+//nolint:nestif
+func (t *txSubmitterImpl) populateGasFromPrevTx(ctx context.Context, transactor *bind.TransactOpts, prevTx *types.Transaction, currentDynamic bool) {
+	if prevTx == nil {
+		return
+	}
+
+	_, span := t.metrics.Tracer().Start(ctx, "submitter.populateGasFromPrevTx")
+
+	defer func() {
+		span.SetAttributes(
+			attribute.String("gas_price", bigPtrToString(transactor.GasPrice)),
+			attribute.String("gas_fee_cap", bigPtrToString(transactor.GasFeeCap)),
+			attribute.String("gas_tip_cap", bigPtrToString(transactor.GasTipCap)),
+		)
+		metrics.EndSpan(span)
+	}()
+
+	prevDynamic := prevTx.Type() == types.DynamicFeeTxType
+	if currentDynamic {
+		if prevDynamic {
+			transactor.GasFeeCap = core.CopyBigInt(prevTx.GasFeeCap())
+			transactor.GasTipCap = core.CopyBigInt(prevTx.GasTipCap())
+		} else {
+			transactor.GasFeeCap = core.CopyBigInt(prevTx.GasPrice())
+			transactor.GasTipCap = core.CopyBigInt(prevTx.GasPrice())
 		}
-		if transactor.GasFeeCap != nil && maxPrice.Cmp(transactor.GasFeeCap) < 0 {
-			transactor.GasFeeCap = maxPrice
+	} else {
+		if prevDynamic {
+			transactor.GasPrice = core.CopyBigInt(prevTx.GasFeeCap())
+		} else {
+			transactor.GasPrice = core.CopyBigInt(prevTx.GasPrice())
 		}
+	}
+}
```

Comment:
Why aren't we setting the bumped values here? For me it makes sense to have "populate from old" and "bump by percentage" in the same function, as these always go together and refer to the same thing

Reply:
Yeah this makes sense, will simplify the code a bit. Will rename `populateGasFromPrevTx` since it will also be doing the bumping

Reply:
Done here: https://github.com/synapsecns/sanguine/pull/2462/commits/d245a735186ad8f6d12aeba5950b98198c5ca2bf